lib/events/pgevents/pgevents.go

```diff
@@ -0,0 +1,584 @@
+// (with some additional query params for our own options).
+func (c *Config) SetFromURL(u *url.URL) error {
+	if u == nil {
+		return nil
+	}
+
+	poolConfig, err := pgxpool.ParseConfig(u.String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	c.PoolConfig = poolConfig
+
+	popParam := func(k string) (string, bool) {
+		if v, ok := poolConfig.ConnConfig.RuntimeParams[k]; ok {
+			delete(poolConfig.ConnConfig.RuntimeParams, k)
+			return v, true
+		}
+		return "", false
+	}
+
+	if s, ok := popParam(authModeParam); ok {
+		c.AuthMode = AuthMode(s)
+	}
+
+	if s, ok := popParam(azureClientIDParam); ok {
+		c.AzureClientID = s
+	}
+
+	if s, ok := popParam(disableCleanupParam); ok {
+		b, err := strconv.ParseBool(s)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		c.DisableCleanup = b
+	}
+
+	if s, ok := popParam(cleanupIntervalParam); ok {
+		d, err := time.ParseDuration(s)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		c.CleanupInterval = d
+	}
+
+	if s, ok := popParam(retentionPeriodParam); ok {
+		d, err := time.ParseDuration(s)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		c.RetentionPeriod = d
+	}
+
+	return nil
+}
+
+// CheckAndSetDefaults checks if the Config is valid, setting default parameters
+// where they're unset. PoolConfig is only checked for presence.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.PoolConfig == nil || c.PoolConfig.ConnConfig == nil {
+		return trace.BadParameter("missing pool config")
+	}
+
+	if err := c.AuthMode.Check(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if c.AzureClientID != "" && c.AuthMode != AzureADAuth {
+		return trace.BadParameter("azure client ID requires azure auth mode")
+	}
+
+	if c.RetentionPeriod < 0 {
+		return trace.BadParameter("retention period must be non-negative")
+	}
+	if c.RetentionPeriod == 0 {
+		c.RetentionPeriod = defaultRetentionPeriod
+	}
+
+	if c.CleanupInterval < 0 {
+		return trace.BadParameter("cleanup interval must be non-negative")
+	}
+	if c.CleanupInterval == 0 {
+		c.CleanupInterval = defaultCleanupInterval
+	}
+
+	if c.Log == nil {
+		c.Log = logrus.WithField(trace.Component, componentName)
+	}
+
+	return nil
+}
+
+// Returns a new Log given a Config. Starts a background cleanup task unless
+// disabled in the Config.
+func New(ctx context.Context, cfg Config) (*Log, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if cfg.AuthMode == AzureADAuth {
+		bc, err := pgcommon.AzureBeforeConnect(cfg.AzureClientID, cfg.Log)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		cfg.PoolConfig.BeforeConnect = bc
+	}
+
+	cfg.Log.Info("Setting up events backend.")
+
+	pgcommon.TryEnsureDatabase(ctx, cfg.PoolConfig, cfg.Log)
+
+	pool, err := pgxpool.NewWithConfig(ctx, cfg.PoolConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	periodicCtx, cancel := context.WithCancel(context.Background())
+	l := &Log{
+		log:    cfg.Log,
+		pool:   pool,
+		cancel: cancel,
+	}
```

Comment:
`periodicCtx` may be unused with disabled cleanup, but we should clean it up anyway, so it shouldn't leak or be an actual drain on resources.

Reply:
If we only create the context when the cleanup is not disabled we must check that `cancel` is non-nil in `Close`; I found it cleaner to just always have it.